### PR TITLE
Implement PSR-17

### PR DESCRIPTION
**Pull request by sagikazarmark**
Replaces #210

### CODE REVIEW

File: src/Factory/HttpFactory.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GuzzleHttp\Psr7\Factory;
```
Comment (Nyholm) [inline src/Factory/HttpFactory.php]: Maybe put it in `GuzzleHttp\Psr7`

File: src/Factory/HttpFactory.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\StreamInterface;
+use Psr\Http\Message\UploadedFileInterface;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Implements all of the PSR-17 interfaces.
+ */
+final class HttpFactory implements
```
Comment (Nyholm) [inline src/Factory/HttpFactory.php]: Im okey with this name.

**Comment by sagikazarmark** [comment]
We can leave the name `HttpFactory` for now, although I'm still not 100% convinced it's the best name.

File: src/HttpFactory.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        return \GuzzleHttp\Psr7\stream_for($resource);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createServerRequest(string $method, $uri, array $serverParams = []): ServerRequestInterface
+    {
+        if (empty($method)) {
```
Comment (gmponos) [inline src/HttpFactory.php]: Shouldn't this be `$method === ''`?

I made the same comment to @Nyholm  [here ](https://github.com/guzzle/psr7/pull/210/files#r230211710) do you believe that this is also `premature optimization`... 

(first time that I heard this term by you @sagikazarmark... good to know 👍 )

Comment (sagikazarmark) [reply src/HttpFactory.php]: Honestly, I don't think it matters that much. You are free to create a PR if it bugs your OCD, but yeah, I would classify this as premature optimization 😄

**Comment by barryvdh** [comment]
Shouldn't this also `provide` the message-factory implementation in composer.json?

File: src/HttpFactory.php (hunk 3 above)
Comment (Tobion) [reply src/HttpFactory.php]: That logic seems to contradict what psr-17 says

> Note that server parameters are taken precisely as given - no parsing/processing
>  of the given values is performed. In particular, no attempt is made to
 >  determine the HTTP method or URI, which must be provided explicitly.

**Pull request merged**